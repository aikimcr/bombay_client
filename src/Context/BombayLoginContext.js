import { createContext } from "react";

const BombayLoginContext = createContext({
  loggedIn: false,
  showLoginForm: false,
  setLoggedIn: null,
  setShowLogin: null,
});

export default BombayLoginContext;
