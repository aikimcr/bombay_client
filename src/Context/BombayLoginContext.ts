import { createContext, Dispatch } from "react";

interface BombayLoginContextValues {
  loggedIn: boolean;
  setLoggedIn?: Dispatch<boolean>;
}

export const BombayLoginContext = createContext({
  loggedIn: false,
  setLoggedIn: null,
});

export default BombayLoginContext;
