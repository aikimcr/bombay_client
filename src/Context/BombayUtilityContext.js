import { createContext } from 'react';

const BombayUtilityContext = createContext({});

export default BombayUtilityContext;
