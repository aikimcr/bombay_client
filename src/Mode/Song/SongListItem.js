import React, { useState } from 'react';

import './SongList.scss';

import Song from './Song';
import { FormModal } from '../../Modal/FormModal';

function SongListItem(props) {
    const [showEdit, setShowEdit] = useState(false);
    const [displayName, setDisplayName] = useState(props.song.get('name'));

    async function updateSong(modelDef) {
        await props.song.set(modelDef).save();
        setDisplayName(props.song.get('name'));
    }

    return (
        <React.Fragment>
            <li className="card" onClick={() => setShowEdit(true)}>
                <div className='header'>Song</div>
                <div className='name'>{displayName}</div>
                <div className='details'>
                    <>
                        <div className='label'>Artist</div>
                        <div className='text'>{props.song.artist().get('name')}</div>
                    </>
                    <>
                        <div className='label'>Key</div>
                        <div className='text'>{props.song.get('key_signature')}</div>
                    </>
                    <>
                        <div className='label'>Tempo</div>
                        <div className='text'>{props.song.get('tempo')}</div>
                    </>
                    <>
                        <div className='label'>Lyrics</div>
                        <div className='longText'>{props.song.get('lyrics')}</div>
                    </>
                </div>
            </li>
            <FormModal 
                title="Edit Song"
                onClose={() => setShowEdit(false)}
                onSubmit={updateSong}
                open={showEdit}
            >
                <Song song={props.song} />
            </FormModal>
        </React.Fragment>
    )
}

export default SongListItem;