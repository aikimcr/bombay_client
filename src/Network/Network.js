// Manage XHR request calls.  This is the low-level stuff.
// import serverConfig from './serverConfig';
const serverConfig = {};

const serverConfigKeys = {
  serverProtocol: 'REACT_APP_SERVER_PROTOCOL',
  serverHost: 'REACT_APP_SERVER_HOST',
  serverBasePath: 'REACT_APP_SERVER_BASE_PATH',
  serverPort: 'REACT_APP_SERVER_PORT',
}

for (const key in serverConfigKeys) {
  if (process.env.hasOwnProperty(serverConfigKeys[key])) {
    const value = process.env[serverConfigKeys[key]];

    switch(value) {
      case 'null':
      case 'none':
      case 'empty':
        serverConfig[key] = null;
        break;

      default:
        serverConfig[key] = value;
    }
  }
}

function setConfigOption(key, defaultConfig) {
  return serverConfig.hasOwnProperty(key) ? serverConfig[key] : defaultConfig;
}

export const serverProtocol = setConfigOption('serverProtocol', 'http');
export const serverHost = setConfigOption('serverHost', 'localhost');
export const serverBasePath = setConfigOption('serverBasePath', '');
export const serverPort = setConfigOption('serverPort', 2001);

console.log(`Server Base URL: ${prepareURLFromArgs('')}`);

function getStandardHeaders(includeContentType = true) {
  const result = {};
  if (includeContentType) result['content-type'] = 'application/json';

  const token = localStorage.getItem('jwttoken')

  if (token) {
    result.Authorization = `Bearer ${token}`;
  }

  return result;
}

export function normalizeAndJoinPath(...pathParts) {
  const newPath = pathParts.reduce((memo, part) => {
    if (part === undefined) {
      throw new Error(`Unable to normalize path with parts: "${pathParts.join(', ')}"`);
    }

    let newPart = memo + '/' + part;
    newPart = newPart.replace(/\/\/+/g, '/');

    if (part !== '/') {
      if (newPart.length > 1) {
        newPart = newPart.replace(/\/+$/, '');
      }
    }

    return newPart;
  }, '/');

  return newPath;
}

export function buildURL(args = {}) {
  let { hostname, path, protocol, port } = {
    hostname: serverHost,
    path: '/',
    protocol: serverProtocol,
    port: serverPort,
    ...args,
  }

  hostname = hostname.replace(/^\/+/, '').replace(/\/+$/, '');

  if (hostname.length === 0) {
    hostname = serverHost;
  }

  let result = `${protocol}://${hostname}`;

  if (port) {
    result += `:${port}`;
  }

  if (Array.isArray(path)) {
    path = normalizeAndJoinPath(...path);
  }

  result += path;

  return result;
}

export function prepareURLFromArgs(path, query) {
  const requestUrl = new URL(buildURL());
  requestUrl.pathname = normalizeAndJoinPath(serverBasePath, path);

  for (const param in query) {
    requestUrl.searchParams.set(param, query[param]);
  }

  return requestUrl;
}

async function decodeResponse(response) {
  if (response.ok) {
    const text = await response.text();
    let result = text;

    try {
      result = JSON.parse(text);
    } catch(err) {
      result = text;
    }

    return result;
  }

  return Promise.reject({ status: response.status, message: response.statusText });
}

export async function getFromURLString(urlString) {
  const response = await fetch(urlString, {
    mode: 'cors',
    credentials: 'include',
    headers: getStandardHeaders(false),
  });

  return decodeResponse(response);
}

function buildJSON(body) {
  const sendBody = { ...body };
  delete sendBody.id;
  delete sendBody.url;
  return JSON.stringify(sendBody);
}

export async function postToURLString(urlString, body) {
  const sendJSON = buildJSON(body);
  const response = await fetch(urlString, {
    method: 'POST',
    headers: getStandardHeaders(),
    body: sendJSON,
    mode: 'cors',
    credentials: 'include',
  });

  return decodeResponse(response);
}

export async function putToURLString(urlString, body) {
  const sendJSON = buildJSON(body);
  const response = await fetch(urlString, {
    method: 'PUT',
    headers: getStandardHeaders(),
    body: sendJSON,
    mode: 'cors',
    credentials: 'include',
  });

  return decodeResponse(response);
}