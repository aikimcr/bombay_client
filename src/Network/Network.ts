import { omit } from 'lodash';

// Manage XHR request calls.  This is the low-level stuff.
// import serverConfig from './serverConfig';
export const defaultAPIServer: string = process.env.SERVER;
export const defaultAPIBasePath: string = process.env.SERVER_BASE_PATH;

const getStandardHeaders = (includeContentType = true) => {
  const result: HeadersInit = {};
  if (includeContentType) {
    result['content-type'] = 'application/json';
  }

  const token = localStorage.getItem('jwttoken');

  if (token) {
    result.Authorization = `Bearer ${token}`;
  }

  return result;
};

export const normalizeAndJoinPath = (...pathList: string[]): string => {
  const newPath = pathList.reduce((memo, part) => {
    if (part === undefined) {
      throw new Error(
        `Unable to normalize path with parts: "${pathList.join(', ')}"`,
      );
    }

    let newPart = memo + '/' + part;
    newPart = newPart.replace(/\/\/+/g, '/');

    if (part !== '/') {
      if (newPart.length > 1) {
        newPart = newPart.replace(/\/+$/, '');
      }
    }

    return newPart;
  }, '/');

  return newPath;
};

interface buildUrlProps {
  server?: string;
  basePath?: string;
  applicationPaths?: string[];
}

export const buildURL = ({
  server = defaultAPIServer,
  basePath = defaultAPIBasePath,
  applicationPaths = [],
}: buildUrlProps = {}) => {
  const allPaths = [...applicationPaths];

  if (!server || server === 'default') {
    server = 'http://localhost:2001';
  }

  if (basePath && basePath !== 'default') {
    allPaths.unshift(basePath);
  }

  const result: URL = new URL(server);
  result.pathname = normalizeAndJoinPath(...allPaths);
  return result;
};

export type NetworkURLQueryType = Record<string, string | number | symbol>;

export const prepareURLFromArgs = (
  pathList: string[] = [],
  query: NetworkURLQueryType = {},
) => {
  const requestUrl = buildURL({ applicationPaths: pathList });

  for (const param in query) {
    requestUrl.searchParams.set(param, query[param].toString());
  }

  return requestUrl;
};

const decodeResponse = async (
  response: Awaited<ReturnType<typeof fetch>>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): Promise<Record<string, any>> => {
  if (response.ok) {
    const text = await response.text();
    let result = { body: text };

    try {
      result = JSON.parse(text);
    } catch (err) {
      result = { body: text };
    }

    return result;
  }

  return Promise.reject({
    status: response.status,
    message: response.statusText,
  });
};

export const getFromURLString = async (
  urlString: string,
): ReturnType<typeof decodeResponse> => {
  const response = await fetch(urlString, {
    mode: 'cors',
    credentials: 'include',
    headers: getStandardHeaders(false),
  });

  return decodeResponse(response);
};

interface buildJSONProps {
  id?: string | number;
  url?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any;
}

const buildJSON = (body: buildJSONProps): string => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const sendBody: Record<string, any> = {
    ...omit(body, ['id', 'url']),
  };
  return JSON.stringify(sendBody);
};

export const postToURLString = async (
  urlString: string,
  body: buildJSONProps,
) => {
  const sendJSON = buildJSON(body);
  const response = await fetch(urlString, {
    method: 'POST',
    headers: getStandardHeaders(),
    body: sendJSON,
    mode: 'cors',
    credentials: 'include',
  });

  return decodeResponse(response);
};

export const putToURLString = async (
  urlString: string,
  body: buildJSONProps,
) => {
  const sendJSON = buildJSON(body);
  const response = await fetch(urlString, {
    method: 'PUT',
    headers: getStandardHeaders(),
    body: sendJSON,
    mode: 'cors',
    credentials: 'include',
  });

  return decodeResponse(response);
};
