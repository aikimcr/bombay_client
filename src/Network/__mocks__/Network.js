const RealNetwork = jest.requireActual('../Network');

export const serverProtocol = 'https';
export const serverHost = 'fake';
export const serverBasePath = 'xyzzy';
export const serverPort = null;

export const normalizeAndJoinPath = RealNetwork.normalizeAndJoinPath;

export function buildURL (args = {}) {
    let { hostname, path, protocol, port } = {
        hostname: serverHost,
        path: '/',
        protocol: serverProtocol,
        port: serverPort,
        ...args,
    }

    return RealNetwork.buildURL(args);
}

// ToDo Find a way to do this without having to copy the entire real implementation.
export function prepareURLFromArgs(path, query) {
    const requestUrl = new URL(buildURL());
    requestUrl.pathname = normalizeAndJoinPath(serverBasePath, path);

    for (const param in query) {
        requestUrl.searchParams.set(param, query[param]);
    }

    return requestUrl;
}


export function getFromURLString(url, query) {
    throw new Error(`getFromURLString has not been mocked properly ${url} ${JSON.stringify(query)}`);
}

export async function postToURLString(urlString, body) {
    throw new Error(`postToURLString has not been mocked properly ${url} ${JSON.stringify(body)}`);
}

export async function putToURLString(urlString, body) {
    throw new Error(`putToURLString has not been mocked properly ${url} ${JSON.stringify(body)}`);
}

// Test Utilities not found in the original
let mockPromise;
let mockResolve;
let mockReject;

export function _setupMockPromise() {
    [mockPromise, mockResolve, mockReject] = makeResolvablePromise();
    return {
        promise: mockPromise,
        resolve: mockResolve,
        reject: mockReject,
    };
}

export function _setupMocks() {
    _setupMockPromise();

    this.getFromURLString = jest.fn((urlString, query) => {
        expect(mockPromise).toBeDefined();
        return mockPromise;
    });

    this.postToURLString = jest.fn((urlString, body) => {
        expect(mockPromise).toBeDefined();
        return mockPromise;
    });

    this.putToURLString = jest.fn((urlString, body) => {
        expect(mockPromise).toBeDefined();
        return mockPromise;
    });

    return {
        resolve: mockResolve,
        reject: mockReject,
        promise: mockPromise,
    };
}