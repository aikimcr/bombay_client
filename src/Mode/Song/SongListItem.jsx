import React, { useState } from "react";

import PropTypes from "prop-types";

import "./SongList.scss";

import { Song } from "./Song";
import FormModal from "../../Modal/FormModal";

export const SongListItem = ({ song }) => {
  const [showEdit, setShowEdit] = useState(false);
  const [displayName, setDisplayName] = useState(song.get("name"));
  const [artistName, setArtistName] = useState(song.artist().get("name"));

  async function updateSong(modelDef) {
    await song.set(modelDef).save();
    setDisplayName(song.get("name"));
    setArtistName(song.artist().get("name"));
  }

  return (
    <>
      <li
        className="card wide"
        data-testid="song-list-card"
        onClick={() => setShowEdit(true)}
      >
        <div className="header">Song</div>
        <div className="name">{displayName}</div>
        <div className="details">
          <>
            <div className="label">Artist</div>
            <div className="text">{artistName}</div>
          </>
          <>
            <div className="label">Key</div>
            <div className="text">{song.get("key_signature")}</div>
          </>
          <>
            <div className="label">Tempo</div>
            <div className="text">{song.get("tempo")}</div>
          </>
          <>
            <div className="label">Lyrics</div>
            <div className="long-text">
              <pre>{song.get("lyrics")}</pre>
            </div>
          </>
        </div>
      </li>
      <FormModal
        title="Edit Song"
        onClose={() => setShowEdit(false)}
        onSubmit={updateSong}
        open={showEdit}
      >
        <Song song={song} />
      </FormModal>
    </>
  );
};

SongListItem.propTypes = {
  song: PropTypes.object,
};

export default SongListItem;
