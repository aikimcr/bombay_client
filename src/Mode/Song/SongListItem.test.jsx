import { useState } from 'react';
import { act, fireEvent, render, screen } from '@testing-library/react';

jest.mock('./Song');
jest.mock('../../Modal/FormModal', () => {
  const originalModule = jest.requireActual('../../Modal/FormModal');

  return {
    __esModule: true,
    ...originalModule,
    default: () => <div data-testid="edit-song-modal">Form Modal</div>,
  };
});

import {
  mockLoginStatus,
  mockRefreshToken,
  mockLogin,
  mockLogout,
} from '../../Network/testing';

jest.mock('../../Network/Login', () => {
  const originalModule = jest.requireActual('../../Network/Login');

  return {
    __esModule: true,
    ...originalModule,
    loginStatus: mockLoginStatus,
    refreshToken: mockRefreshToken,
    login: mockLogin,
    logout: mockLogout,
  };
});

import * as Network from '../../Network/Network';

import {
  makeFetchBodyAndModels,
  makeAModel,
} from '../../testHelpers/modelTools';

import BombayLoginContext from '../../Context/BombayLoginContext';
import BombayUtilityContext from '../../Context/BombayUtilityContext';

import { SongListItem } from './SongListItem';

jest.useFakeTimers();

const mockUtility = {
  getBootstrap: () => {
    return {
      keySignatures: ['A', 'B', 'C', 'D', 'E', 'F', 'G'],
    };
  },
};

const renderWrapper = ({ children }) => {
  const [loginState, setLoginState] = useState({
    loggedIn: true,
    showLoginForm: false,
  });

  return (
    <BombayLoginContext.Provider value={loginState}>
      <BombayUtilityContext.Provider value={mockUtility}>
        {children}
      </BombayUtilityContext.Provider>
    </BombayLoginContext.Provider>
  );
};

function setupLogin(loggedIn = true) {
  mockLoginStatus.mockResolvedValue(loggedIn);
}

const makeTestSong = () => {
  const [_artistModelDef, artistModel] = makeAModel('artist', (def) => {
    def.name = 'Billy Joel';
  });
  const [_modelDef, model] = makeAModel('song', (def) => {
    def.name = 'Piano Man';
    def.artist = artistModel;
  });

  return model;
};

it('Component should match snapshot', async () => {
  const model = makeTestSong();

  const { asFragment } = render(<SongListItem song={model} />, {
    wrapper: renderWrapper,
  });

  expect(asFragment()).toMatchSnapshot();
});

it('should open the editor modal', async () => {
  setupLogin();

  const model = makeTestSong();
  render(<SongListItem song={model} />, {
    wrapper: renderWrapper,
  });

  const songCard = screen.getByTestId('song-list-card');

  await act(async () => {
    fireEvent.click(songCard);
  });

  expect(screen.getByTestId('edit-song-modal')).toBeInTheDocument();
});

// Move these tests to the SongForm component
it.skip('should save changes to the model', async () => {
  setupLogin();

  const [mockGetPromise, mockGetResolve] = makeResolvablePromise();
  const [mockPutPromise, mockPutResolve] = makeResolvablePromise();
  // The mock is not recognized unless it is done this way.
  Network.getFromURLString = jest.fn((url) => {
    return mockGetPromise;
  });

  Network.putToURLString = jest.fn((url, body) => {
    return mockPutPromise;
  });

  const [fetchBody, models] = makeFetchBodyAndModels(10, {});

  await act(async () => {
    mockGetResolve(fetchBody);
  });

  const model = makeTestSong();
  const songModelDef = { ...modelDef };
  delete songModelDef.artist;

  const result = render(<SongListItem song={model} />, {
    wrapper: renderWrapper,
  });

  const index = result.container;

  await act(async () => {
    index.firstChild.click();
  });

  const modalRoot = document.getElementById('modal-root');
  expect(modalRoot.childElementCount).toEqual(1);

  const submitButton = modalRoot.querySelector('[type="submit"]');

  await changeInput(
    modalRoot.querySelector('[data-targetfield="name"'),
    'input',
    'Herkimer',
    250,
  );

  act(() => {
    submitButton.click();
  });

  await act(async () => {
    mockPutResolve({ ...modelDef, name: 'Herkimer' });
  });

  expect(Network.putToURLString).toBeCalledTimes(1);
  expect(Network.putToURLString).toBeCalledWith(modelDef.url, {
    ...songModelDef,
    name: 'Herkimer',
  });
});

it.skip('should update the artist name', async () => {
  setupLogin();

  const [mockGetPromise, mockGetResolve] = makeResolvablePromise();
  const [mockPutPromise, mockPutResolve] = makeResolvablePromise();
  // The mock is not recognized unless it is done this way.
  Network.getFromURLString = jest.fn((url) => {
    return mockGetPromise;
  });

  Network.putToURLString = jest.fn((url, body) => {
    return mockPutPromise;
  });

  const [modelDef, songModel] = makeAModel('song');
  const oldArtistDef = modelDef.artist;
  const songModelDef = { ...modelDef };
  delete songModelDef.artist;

  const result = render(<SongListItem song={songModel} />, {
    wrapper: renderWrapper,
  });

  const index = result.container;

  await act(async () => {
    index.firstChild.click();
  });

  const modalRoot = document.getElementById('modal-root');
  expect(modalRoot.childElementCount).toEqual(1);

  const submitButton = modalRoot.querySelector('[type="submit"]');
  const editButton = modalRoot.querySelector('.picker-button button');
  expect(editButton.textContent).toEqual(oldArtistDef.name);

  const [newArtistDef, newArtistModel] = makeAModel('artist');
  const fetchBody = {
    data: [oldArtistDef, newArtistDef],
    url: 'xyzzy',
  };

  await act(async () => {
    editButton.click();
    mockGetResolve(fetchBody);
  });

  const newArtistLine = modalRoot.querySelector('li.picker-item:last-child');
  expect(newArtistLine).toHaveTextContent(newArtistDef.name);

  songModelDef.artist_id = newArtistDef.id;

  await act(async () => {
    newArtistLine.click();
  });

  expect(editButton).toHaveTextContent(newArtistDef.name);

  await act(async () => {
    submitButton.click();
    mockPutResolve({ ...songModelDef, artist: newArtistDef });
  });

  expect(Network.putToURLString).toBeCalledTimes(1);
  expect(Network.putToURLString).toBeCalledWith(modelDef.url, songModelDef);
  expect(songModel.artist().toJSON()).toEqual(newArtistDef);
});
