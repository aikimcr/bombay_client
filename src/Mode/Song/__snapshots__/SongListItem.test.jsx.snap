// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Component should match snapshot 1`] = `
<DocumentFragment>
  <li
    class="card"
    data-testid="song-list-card"
  >
    <div
      class="header"
    >
      Song
    </div>
    <div
      class="name"
    >
      Piano Man
    </div>
    <div
      class="details"
    >
      <div
        class="label"
      >
        Artist
      </div>
      <div
        class="text"
      />
      <div
        class="label"
      >
        Key
      </div>
      <div
        class="text"
      />
      <div
        class="label"
      >
        Tempo
      </div>
      <div
        class="text"
      />
      <div
        class="label"
      >
        Lyrics
      </div>
      <div
        class="long-text"
      >
        <pre />
      </div>
    </div>
  </li>
  <div
    data-testid="edit-song-modal"
  >
    Form Modal
  </div>
</DocumentFragment>
`;
